Update Readme.md (#1)
DevExpress-Examples/asp-net-web-forms-gridview-select-deselect-all-rows-in-a-group: Readme.md

<!-- default badges list -->
[![](https://img.shields.io/badge/Open_in_DevExpress_Support_Center-FF7200?style=flat-square&logo=DevExpress&logoColor=white)](https://supportcenter.devexpress.com/ticket/details/E1760)
[![](https://img.shields.io/badge/📖_How_to_use_DevExpress_Examples-e9f6fc?style=flat-square)](https://docs.devexpress.com/GeneralInformation/403183)
<!-- default badges end -->

> # Grid View for ASP.NET Web Forms- How to select/deselect all rows in a group when the control is grouped by a column

Reviewer: A typo.

```suggestion
# Grid View for ASP.NET Web Forms - How to select/deselect all rows in a group when the control is grouped by a column
```

> # Grid View for ASP.NET Web Forms - How to select/deselect all rows in a group when the control is grouped by a column

Reviewer 2: ```suggestion
# Grid View for ASP.NET Web Forms - How to select/deselect all rows in a group when a control is grouped by columns
```

Author: The technique described in this example works only when the Grid is grouped by one column. Also, when I say "the control", I mean "the Grid View control", that's why I use the definite article.
Maybe we can say the following?

```suggestion
# Grid View for ASP.NET Web Forms - How to select/deselect all rows in a group when the control is grouped by one column
```

Reviewer: Hello Armina,

Please excuse me for interfering, but I should have noticed this in the first place. One groups data, not the control itself. Regarding one column -- yes, this is the important part of the described scenario. I hope the following change makes sense:

```suggestion
# Grid View for ASP.NET Web Forms - How to select/deselect all rows in a group when data is grouped by one column
```

Reviewer 2: yes

Reviewer: Thank you!

<!-- run online -->
**[[Run Online]](https://codecentral.devexpress.com/e1760/)**
<!-- run online end -->

> This example demonstrates how to allow users to add/remove all rows in a group to/from selection. Note that this approach works only when the [Grid View](https://docs.devexpress.com/AspNet/5823/components/grid-view?p=netframework)'s data is grouped by a single column.

Reviewer: > Note that this **approach** works only when the [Grid View](https://docs.devexpress.com/AspNet/5823/components/grid-view?p=netframework)'s data is grouped by a single column.

Approach. Usage of this word is being discouraged of late. I prefer "technique" instead.

Is there a specific reason to use the following form?

> the Grid View's data

Instead of "Grid View data"

> is grouped by a **single** column.

Tautology. "A" already means "single", actually. Therefore, "a single" makes extra emphasis on the aspect of singularity. I understand that "is grouped by a column" sounds somewhat strange, and that's what caused you to add "single". I recommend replacing "a single" with "one" instead because it describes our scenario.

```suggestion
This example demonstrates how to allow users to add/remove all rows in a group to/from selection. Note that this technique works only when [Grid View](https://docs.devexpress.com/AspNet/5823/components/grid-view?p=netframework) data is grouped by one column.
```

If you have any questions or doubts on this change - please let me know.

![Select/Deselect a Group](select-deselect.gif)

## Overview

> Follow the steps below to allow users to to select/deselect all rows in a group:

Reviewer: I remove slash between operations wherever it is possible because / can be read as follows:

1. AND
2. OR
3. AND/OR

I understand there are scenarios when the alternative would have been way worse (see above). However, this is not the case, therefore, if we can add certainty here, we definitely should:

```suggestion
Follow the steps below to allow users to to select or deselect all rows in a group:
```

    ```cs
    protected void Grid_HtmlRowPrepared(object sender, ASPxGridViewTableRowEventArgs e) {
        if(e.RowType == GridViewRowType.Group) {
            ASPxCheckBox checkBox = Grid.FindGroupRowTemplateControl(e.VisibleIndex, "checkBox") as ASPxCheckBox;
            if(checkBox != null) {
                checkBox.ClientSideEvents.CheckedChanged = string.Format("function(s, e){{ Grid.PerformCallback('{0};' + s.GetChecked()); }}", e.VisibleIndex);
                checkBox.Checked = GetChecked(e.VisibleIndex);
            }
        }
    }

    protected bool GetChecked(int visibleIndex) {
        for(int i = 0; i < Grid.GetChildRowCount(visibleIndex); i++) {
            bool isRowSelected = Grid.Selection.IsRowSelectedByKey(Grid.GetChildDataRow(visibleIndex, i)["ProductID"]); 
            if(!isRowSelected)
                return false;
        }
        return true;
    }
    ```

> 4. Handle the Grid View's [CustomCallback](https://docs.devexpress.com/AspNet/DevExpress.Web.ASPxGridView.CustomCallback?p=netframework) event to process the callback. Call the Grid View's [GetChildDataRow](https://docs.devexpress.com/AspNet/DevExpress.Web.ASPxGridView.GetChildDataRow(System.Int32-System.Int32)?p=netframework) method to get a data row that belongs to the current group. Pass the data row's key and the the check state of the checkbox to the [SetSelectionByKey](https://docs.devexpress.com/AspNet/DevExpress.Web.Data.WebDataSelection.SetSelectionByKey(System.Object-System.Boolean)?p=netframework) method to select or deselect this row:

Reviewer: > the the check state

Another typo. As you can see, not all typos are highlighted by a spellchecker. Such typos most often come from lengthy editing sessions of a complex sentence or paragraph. I recommend taking another look once your work is complete.

```suggestion
4. Handle the Grid View's [CustomCallback](https://docs.devexpress.com/AspNet/DevExpress.Web.ASPxGridView.CustomCallback?p=netframework) event to process the callback. Call the Grid View's [GetChildDataRow](https://docs.devexpress.com/AspNet/DevExpress.Web.ASPxGridView.GetChildDataRow(System.Int32-System.Int32)?p=netframework) method to get a data row that belongs to the current group. Pass the data row's key and the check state of the checkbox to the [SetSelectionByKey](https://docs.devexpress.com/AspNet/DevExpress.Web.Data.WebDataSelection.SetSelectionByKey(System.Object-System.Boolean)?p=netframework) method to select or deselect this row:
```

    ```cs
    protected void Grid_CustomCallback(object sender, ASPxGridViewCustomCallbackEventArgs e) {
        string[] parameters = e.Parameters.Split(';');
        int index = int.Parse(parameters[0]);
        bool isGroupRowSelected = bool.Parse(parameters[1]);
        for(int i = 0; i < Grid.GetChildRowCount(index); i++) {
            DataRow row = Grid.GetChildDataRow(index, i);
            Grid.Selection.SetSelectionByKey(row["ProductID"], isGroupRowSelected);
        }
    }
    ```

## Files to Review

* [Default.aspx](./CS/WebSite/Default.aspx) (VB: [Default.aspx](./VB/WebSite/Default.aspx))
* [Default.aspx.cs](./CS/WebSite/Default.aspx.cs) (VB: [Default.aspx.vb](./VB/WebSite/Default.aspx.vb))

## Documentation

* [Group Data](https://docs.devexpress.com/AspNet/3715/components/grid-view/concepts/group-data)
* [Callbacks](https://docs.devexpress.com/AspNet/402559/common-concepts/callbacks)

## More Examples

> * [Grid View for ASP.NET MVC- How to select/deselect all rows in a group](https://github.com/DevExpress-Examples/gridview-how-to-implement-select-unselect-for-all-rows-in-a-group-row-t362032)

Reviewer: A typo. Do not forget about necessary space characters.

```suggestion
* [Grid View for ASP.NET MVC - How to select/deselect all rows in a group](https://github.com/DevExpress-Examples/gridview-how-to-implement-select-unselect-for-all-rows-in-a-group-row-t362032)
```

> * [Grid Lookup for ASP.NET Web Forms- How to select/deselect all rows in a group](https://github.com/DevExpress-Examples/how-to-implement-select-unselect-for-all-rows-in-a-group-row-in-aspxgridlookup-t299266)

Reviewer: Likewise.

```suggestion
* [Grid Lookup for ASP.NET Web Forms - How to select/deselect all rows in a group](https://github.com/DevExpress-Examples/how-to-implement-select-unselect-for-all-rows-in-a-group-row-in-aspxgridlookup-t299266)
```
